### PR TITLE
Bug: csidriver crash loop / pod terminating forever

### PR DESCRIPTION
**Pull request by aaaaaaaalex**
## Description

Pods which use nodeagent volumes will stay in Terminating phase forever, if their node reboots while they are in the Terminating phase. The chances of this happening increase when a termination grace period is set on a pod whose container only responds to SIGKILL.

Post-boot, the container is not restored and we don't see its nodeagent files appear in the host FS, since it is meant to be terminating. Kube will then invoke our CSI-driver's TearDown routine. The TearDown routine fails because we cannot find the nodeagent files which we assume are mounted (even though we only want to delete them anyway...).

This PR logs "NotExists" FS errors and continues, rather than logging and exiting, to ensure we don't block the TearDown process.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Fix a bug where pods with flexvol/nodeagent volumes would get stuck in the Terminating phase, if, during termination, their node rebooted.
```
## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: pod2daemon/csidriver/driver/node.go [hunk 1]
```diff
@@ -107,7 +109,19 @@ func (ns *nodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnp
...
 	if err != nil {
 		log.WithError(err).Error("Unable to retrieve pod info")
-		return nil, status.Errorf(codes.Internal, "Unable to retrieve pod info: %s", err)
+		// If the pod-info file is missing, it's likely the container volumes were already unmounted.
+		// This can be the case when a node-restart occurs as a pod is terminating: upon reboot, the
+		// pod container dir will not be mounted, but we still receive a CSI call to unmount volumes.
+		if !errors.Is(err, fs.ErrNotExist) {
+			return nil, status.Errorf(codes.Internal, "Unable to retrieve pod info: %s", err)
+		}
+		log.Info("Ignoring ErrNotExist and continuing with unmount")
```
Comment (fasaxc) [inline pod2daemon/csidriver/driver/node.go:118]:
```suggestion
		log.Info("Pod information file is already gone, continuing with unmount.")
```
Might rephrase to be a little more user accessible.  OK, the user might not know what the pod info file is but the log sortof hints that we were trying to get rid of it.

**Review by fasaxc — CHANGES_REQUESTED** [review]
Few minor suggestions, also could do with tests so we don't regress (and to confirm that the real errors actually match the conditions etc).

File: pod2daemon/csidriver/driver/node.go [hunk 2]
```diff
@@ -118,11 +132,18 @@ func (ns *nodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnp
 
 	// Clean up the file storing the pod info.
 	if err = ns.removeCredentialFile(req.VolumeId); err != nil {
-		log.Errorf("Could not remove pod info file at %s/%s", ns.config.NodeAgentCredentialsHomeDir, req.VolumeId)
-		return nil, status.Errorf(codes.Internal, "Could not remove pod info file at %s/%s", ns.config.NodeAgentCredentialsHomeDir, req.VolumeId)
+		log.WithError(err).WithField("file", fmt.Sprintf("%s/%s", ns.config.NodeAgentCredentialsHomeDir, req.VolumeId)).Error("Could not remove pod info file")
+		if !errors.Is(err, fs.ErrNotExist) {
```
[the remark this answers is not in the document]
Comment (aaaaaaaalex) [reply pod2daemon/csidriver/driver/node.go:136]: Did a quick demo and this equivalence check does seem to work.

Looking at the code, there does seem to be aliasing happening.

Our helper functions call the `os` package (`os.Remove` in this case), but errors returned from that function are all of type `PathError`, and the definition of `os.PathError` is:

```
fs.PathError
```

**Review by fasaxc — APPROVED** [review]
LGTM, thanks for making sure test failure actually fails CI the job.  Good catch.

**Pull request merged**